torch/csrc/autograd/engine.cpp

```diff
@@ -804,7 +872,7 @@ auto Engine::start_threads() -> void {
   }
 }
 
-void Engine::add_thread_pool_task(GraphTask *graph_task) {
+void Engine::add_thread_pool_task(std::weak_ptr<GraphTask> graph_task) {
```

Comment:
Surely you're not taking ownership of the weak pointer here?

Reply:
Yes, but we do take ownership before we call `thread_main`. Didn't want to take ownership here since the task could be in the queue for a while.